Courses crud

madsWesth opened the pull request:
## Description

Main goal for this task was to make some reusable components and hooks.

The Table and SearchBar components are meant to be reusable for the entire project. This is done by having them be purely presentational i.e. the arent concerned with functionality and rely on the parent component for that.

Courses table for admin that use the new stuff.

## Changes

### What's new?
- Table components
- SearchBar component
- Pagination component and hook
- Courses table that uses the new stuff

### What's missing?
- Custom width of columns
- useSearch hook that can be used everywhere in the project

## Related Issues

This PR is dependent on this PR https://github.com/Educado-App/educado-frontend/pull/211 and there is some wonky stuff with the test which is the reason why that one hasnt been merged

## Checklist

- [x] Code has been tested locally and passes all relevant tests.
- [x] Documentation has been updated to reflect the changes, if applicable.
- [x] Code follows the established coding style and guidelines of the project.
- [x] All new and existing tests related to the changes have passed.
- [x] Any necessary dependencies or new packages have been properly documented.
- [x] Pull request title and description are clear and descriptive.
- [x] Reviewers have been assigned to the pull request.
- [x] Any potential security implications have been considered and addressed.
- [ ] Performance impact of the changes has been evaluated, if relevant.
- [x] Changes have been approved by Product Owner
- [ ] Code has been analyzed with CodeScene
- [x] I have reviewed the pull request myself

## Screenshots (if applicable)

Pictures:
![image](https://github.com/user-attachments/assets/d7c8325e-4611-4e3c-898f-b2575803b613)
Courses table. Nothing interesting and still missing some things

![Screenshot 2024-11-18 034026](https://github.com/user-attachments/assets/d3fcfcdb-b759-4f41-bb36-6bc86c24451e)
![Screenshot 2024-11-18 034138](https://github.com/user-attachments/assets/39b5ae6c-6cce-4a08-b939-e751e8a9d5be)
Comparison of use. I have put all the "custom" content for the rows in a rows variable for clarity https://github.com/Educado-App/educado-frontend/pull/205/files#diff-6e532a02a1d22b9f37ffd33bf4773ebbd64d0c1a285f73cf451b2dc13e5876aaR70-R103

## If mobile/frontend pull request, what version of the backend is it stable, and running on? 

Stable on dev

madsWesth commented on app/src/components/Table/index.ts:
This file is so all the components can easily be imported in other places. You can see it in action in `CoursesTableAdmin`

madsWesth commented on app/src/components/Table/contexts.ts:
Smart way to pass some information to nested components. In this instance if the `TableCells` are in the header row of the table so they can get the correct styling

Lowkee1g commented on app/src/components/Admin/Institutions/Actions/AddInstitutionsButton.tsx:
```diff
@@ -0,0 +1,151 @@
+        "Adicionado " + res.institutionName + " como nova instituição"
+      );
+    } catch {
+      if (error) {
```
Why have a check for error here? If we end up in the catch it is more than likely because of an error. Now i do not see an else statement here, that would mean that if this in not an error, we might not even know there is an error. Right? 

2 potential solutions: 
- Remove the check
- Reverse the if statement and add an "throw error"

I might also just be tripping 🥲 

>

LazarShockX replied:
Good observation!
```
catch (error) {

}
```

Lowkee1g reviewed (changes requested):
The code in general looks good. Great work with the AdminTable and smart way of splitting the table using an index file. but we do have problems. 
- Icon are pulled from w3.org, make sure this is cleared with PO iara as this is not from the figma.
- Search bar does not work, and will crash the webpage if used in other pages than institution, see pic:
<img width="1710" alt="Skærmbillede 2024-11-23 kl  14 50 03" src="https://github.com/user-attachments/assets/4540b71c-adec-4fc8-adc0-cf94a0cc100d">
- Search does not work
- On Mac, there seems to be a problem with the drop-down looking funny, might not be fixable, see pic:
<img width="203" alt="Skærmbillede 2024-11-23 kl  14 52 10" src="https://github.com/user-attachments/assets/0d4c5d7f-b6ff-4b87-b7a5-6b7977ab9086">
- in a few files the following comment is at the top of the file. This has to be removed... no easy way out of Eslint 💀 
/* eslint-disable @typescript-eslint/strict-boolean-expressions */

madsWesth commented:
@Lowkee1g 
Thanks for the great review mate and sorry for the messiness with git.

Responses to your different point:

#### Icons
This was left over from the git issues should now be fixed, they should still be changed to use material design from `react-icons` which is standard for the project. I'm not gonna change them now since its not a change i made and has already been approved somewhere else.

#### Search crashes
Ive not had crashes on my end but my database looks different to yours since I'm using a local version and not the staging one. I have tried handling it with some null checks, if you could check that it works for you that would be awesome.

#### Search doesn't work
There is definitely something wrong here but I suggest that we leave it for now since its wrong for all tables and not just this new one.
Im gonna make another pull request that generalizes this functionality so it can be used everywhere and i will fix it there ofc. 😄 

#### Drop down looks weird
Fixed on Courses table. When the other two tables are adapted to use the new components they will look the same. I'm not doing that here since that would be a huge change.

#### ESLint
I agree with you that disabling linting rules should really never be done. I did it here since I think the rule is counter productive and causes more errors than it fixes. In react its fairly standard to conditionally render using the logical short circuit on the `&&`(and) operator. Here this rule makes you handle every case individually instead of any false value, the typing for the project is bad so sometimes it asks you to handle only the specific case that typescript identifies, but with the typing being poor in the project sometimes there are more cases that should be handled which cause crashes. Doing the standard falsy check catches all of these which is why i disabled it in the files i added. I have already talked to someone about disabling this but I will also make a pull request that disables this and hear what people think.

If you want me to not disable the rule i will adapt the code 😇

Lowkee1g reviewed (approved):
looks good to me

Lowkee1g commented:
> @Lowkee1g Thanks for the great review mate and sorry for the messiness with git, which should now be fixed.
> 
> Responses to your different point:
> 
> #### Icons
> This was left over from the git issues should now be fixed, they should still be changed to use material design from `react-icons` which is standard for the project. I'm not gonna change them now since its not a change i made and has already been approved somewhere else.
> 
> #### Search crashes
> Ive not had crashes on my end but my database looks different to yours since I'm using a local version and not the staging one. I have tried handling it with some null checks, if you could check that it works for you that would be awesome.
> 
> #### Search doesn't work
> There is definitely something wrong here but I suggest that we leave it for now since its wrong for all tables and not just this new one. Im gonna make another pull request that generalizes this functionality so it can be used everywhere and i will fix it there ofc. 😄
> 
> #### Drop down looks weird
> Fixed on Courses table. When the other two tables are adapted to use the new components they will look the same. I'm not doing that here since that would be a huge change.
> 
> #### ESLint
> I agree with you that disabling linting rules should really never be done. I did it here since I think the rule is counter productive and causes more errors than it fixes. In react its fairly standard to conditionally render using the logical short circuit on the `&&`(and) operator. Here this rule makes you handle every case individually instead of any false value, the typing for the project is bad so sometimes it asks you to handle only the specific case that typescript identifies, but with the typing being poor in the project sometimes there are more cases that should be handled which cause crashes. Doing the standard falsy check catches all of these which is why i disabled it in the files i added. I have already talked to someone about disabling this but I will also make a pull request that disables this and hear what people think.
> 
> If you want me to not disable the rule i will adapt the code 😇

I have not been able to check if the search works but i will later tonight as i am at work at the moment. I didnt know about the eslint issue, so just keep it as is and if someone later on has a problem with it, they can deal with it :). 

If you have the time, you could test it the crash by switching the Database url

I will be home at around 22:00, you can yolo it, or wait
